File: 1036-rotting-oranges/rotting-oranges.cpp

#include <bits/stdc++.h>
using namespace std;
//bfs
class Solution {
    public:
     //find min time required to rot all oranges. 
    int orangesRotting(vector < vector < int >> & grid) {
        if(grid.empty()) return 0;
        int m=grid.size();
        int n=grid[0].size();
        int time=0,total=0,ctr=0;
        queue<pair<int,int>> rotten;
        for(int i=0;i<m;++i){
            for(int j=0;j<n;++j){
                if(grid[i][j]!=0) total++; //total count of fresh & rotten oranges
                if(grid[i][j]==2) rotten.push({i,j}); //rotten oranges(2) pushed in queue
            }
        }
        int dx[4]={-1,0,1,0};
        int dy[4]={0,-1,0,1};
        while(!rotten.empty()){
            int k=rotten.size();
            ctr+=k;
            while(k--){
                int x=rotten.front().first;
                int y=rotten.front().second;
                rotten.pop();
                for(int i=0;i<4;i++){ //4 dirn
                    int neigh_row=x+dx[i];
                    int neigh_col=y+dy[i];
                    if(neigh_row<0|| neigh_row>=m || neigh_col<0 || neigh_col>=n || grid[neigh_row][neigh_col]!=1){//if out of bounds or not a fresh roange, skip
                        continue;
                    }   
                    grid[neigh_row][neigh_col]=2;//rot it
                    rotten.push({neigh_row,neigh_col});
                }
            }
            if(!rotten.empty()) time++;
        }
        return (total==ctr) ? time:-1;
    }
};

// class Solution {
// public:
//     int orangesRotting(vector<vector<int>>& grid) {
//         int m = grid.size();
//         int n = grid[0].size();
//         vector<vector<int>> visited = grid;
//         //queue in which we will fill rotten oranges
//         queue<pair<int, int>> q;
//         int countFreshOrange = 0;
//         for (int i = 0; i < m; i++) {
//             for (int j = 0; j < n; j++) {
//                 if (visited[i][j] == 2) {
//                     q.push({i, j});
//                 }
//                 if (visited[i][j] == 1) {
//                     countFreshOrange++;
//                 }
//             }
//         }
//         //q.empty() means there is no rotten orange in the grid and countFreshOrange = 0 means we will rotten the freshoranges in 0 mins
//         if (countFreshOrange == 0)
//             return 0;
//         if (q.empty())
//             return -1;
//         int time = -1;
//         //four dirns
//         vector<pair<int, int>> dirs = {{1, 0},{-1, 0},{0, -1},{0, 1}};
//         while (!q.empty()) {
//             int size = q.size();
//             while (size--) {
//                 auto [x, y] = q.front();
//                 q.pop();
//                 for (auto [dx, dy] : dirs) {
//                     int i = x + dx;
//                     int j = y + dy;
//                     if (i >= 0 && i < m && j >= 0 && j < n && visited[i][j] == 1) {
//                         visited[i][j] = 2;
//                         countFreshOrange--;
//                         q.push({i, j});
//                     }
//                 }
//             }
//             time++;
//         }
//         if (countFreshOrange == 0)
//             return time;
//         return -1;
//     }
// };

// class Solution {
//     public:
//       //find min time required to rot all oranges. 
//       int orangesRotting(vector < vector < int >> & grid) {
//         int n = grid.size();
//         int m = grid[0].size();
//         // store {{row, column}, time}
//         queue < pair < pair < int, int > , int >> q;
//         int vis[n][m];
//         int cntFresh = 0;
//         for (int i = 0; i < n; i++) {
//           for (int j = 0; j < m; j++) {
//             // if cell contains rotten orange
//             if (grid[i][j] == 2) {
//               q.push({{i, j}, 0}); 
//               // mark as visited (rotten) in visited array
//               vis[i][j] = 2;
//             }
//             // if not rotten
//             else {
//               vis[i][j] = 0;
//             }
//             // count fresh oranges
//             if (grid[i][j] == 1) cntFresh++;
//           }
//         }
//         int tm = 0;
//         // delta row and delta column
//         int drow[] = {-1, 0, +1, 0};
//         int dcol[] = {0, 1, 0, -1}; 
//         int cnt = 0;
//         // bfs traversal (until the queue becomes empty)
//         while (!q.empty()) {
//           int r = q.front().first.first;
//           int c = q.front().first.second;
//           int t = q.front().second;
//           tm = max(tm, t);
//           q.pop();
//           // exactly 4 neighbours 
//           for (int i = 0; i < 4; i++) {
//             // neighbouring row and column
//             int nrow = r + drow[i];
//             int ncol = c + dcol[i];
//             // check for valid cell and 
//             // then for unvisited fresh orange
//             if (nrow >= 0 && nrow < n && ncol >= 0 && ncol < m &&
//               vis[nrow][ncol] == 0 && grid[nrow][ncol] == 1) {
//               // push in queue with timer increased
//               q.push({{nrow, ncol}, t + 1}); 
//               // mark as rotten
//               vis[nrow][ncol] = 2;
//               cnt++;
//             }
//           }
//         }
//         // if all oranges are not rotten
//         if (cnt != cntFresh) return -1;
//         return tm;
//     }
// };

File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q=deque()
        m, n = len(grid), len(grid[0])
        time,fresh=0,0
        for r in range(m):
            for c in range(n):
                if(grid[r][c]==1):
                    fresh+=1
                if(grid[r][c]==2):
                    q.append([r,c])
        dirn=[[0,1],[0,-1],[1,0],[-1,0]]
        while(q and fresh>0):
            for i in range(len(q)):
                r,c=q.popleft()
                for dr,dc in dirn:
                    row,col=dr+r,dc+c
                    # if in bounds and fresh- then rot it
                    if(row<0 or row==len(grid) or col<0 or col==len(grid[0]) or grid[row][col]!=1):
                        continue
                    grid[row][col]=2
                    q.append([row,col])
                    fresh-=1
            time+=1
        return time if fresh==0 else -1
# # Similar Pattern Problems:
# # As Far from Land as Possible - https://leetcode.com/problems/as-far-from-land-as-possible/description/
# # Shortest Distance from All Buildings - https://leetcode.com/problems/shortest-distance-from-all-buildings/description/
# class Solution:
#     def orangesRotting(self, grid: List[List[int]]) -> int:
#         m, n = len(grid), len(grid[0])
#         visited = grid
#         q = collections.deque()
#         freshctr = 0
#         for i in range(m):
#             for j in range(n):
#                 if visited[i][j] == 2:
#                     q.append((i, j))
#                 if visited[i][j] == 1:
#                     freshctr += 1
#         if freshctr == 0:
#             return 0
#         if not q:
#             return -1
#         time = -1
#         dirs = [(1, 0), (-1, 0), (0, -1), (0, 1)]
#         while q:
#             size = len(q)
#             while size > 0:
#                 x, y = q.popleft()
#                 size -= 1
#                 for dx, dy in dirs:
#                     i, j = x + dx, y + dy
#                     if 0 <= i < m and 0 <= j < n and visited[i][j] == 1:
#                         visited[i][j] = 2
#                         freshctr -= 1
#                         q.append((i, j))
#             time += 1
#         if freshctr == 0:
#             return time
#         return -1
# # from collections import deque
# # # T: O(m * n) -> each cell is visited at least once
# # # S: O(m * n) -> in the worst case if all the oranges are rotten they will be added to the queue
# # class Solution:
# #     def orangesRotting(self, grid: List[List[int]]) -> int:
# #         m = len(grid)
# #         if m == 0:  # check if grid is empty
# #             return -1        
# #         n = len(grid[0])
# #         # keep track of fresh oranges
# #         freshctr = 0
# #         # queue with rotten oranges (for BFS)
# #         rotten = deque()        
# #         # visit each cell in the grid
# #         for r in range(m):
# #             for c in range(n):
# #                 if grid[r][c] == 2:
# #                     # add the rotten orange coordinates to the queue
# #                     rotten.append((r, c))
# #                 elif grid[r][c] == 1:
# #                     # update fresh oranges count
# #                     freshctr += 1
# #         time = 0
# #         # If there are rotten oranges in the queue and there are still fresh oranges in the grid keep looping
# #         while rotten and freshctr > 0:
# #             # update the number of time passed
# #             # it is safe to update the time by 1, since we visit oranges level by level in BFS traversal.
# #             time += 1
# #             # process rotten oranges on the current level
# #             for _ in range(len(rotten)):
# #                 x, y = rotten.popleft()
# #                 # visit all the adjacent cells
# #                 for dx, dy in [(1,0), (-1,0), (0,1), (0,-1)]:
# #                     # calculate the coordinates of the adjacent cell
# #                     nrow, ncol = x + dx, y + dy
# #                     # ignore the cell if it is out of the grid boundary
# #                     if nrow < 0 or nrow == m or ncol < 0 or ncol == n:
# #                         continue
# #                     # ignore the cell if it is empty '0' or visited before '2'
# #                     if grid[nrow][ncol] == 0 or grid[nrow][ncol] == 2:
# #                         continue
# #                     # update the fresh oranges count
# #                     freshctr -= 1
# #                     # mark the current fresh orange as rotten
# #                     grid[nrow][ncol] = 2
# #                     # add the current rotten to the queue
# #                     rotten.append((nrow, ncol))
# #         # return the number of time taken to make all the fresh oranges to be rotten
# #         # return -1 if there are fresh oranges left in the grid (there were no adjacent rotten oranges to make them rotten)
# #         return time if freshctr == 0 else -1
